File: exercise/ts/day22/src/eid/eid.ts
import {Either, isLeft, left, right, chain} from 'fp-ts/Either';
import {Sex} from "./sex";
import {Year} from "./year";
import {SerialNumber} from "./serialNumber";
import {ParsingException} from "./parsingException";
import {pipe} from "fp-ts/function";

export class EID {
    constructor(
        private readonly sex: Sex,
        private readonly year: Year,
        private readonly serialNumber: SerialNumber
    ) {
    }

    static parse(potentialEID: string): Either<ParsingException, EID> {
        return pipe(
            Sex.parse(potentialEID[0]),
            chain(sex =>
                pipe(
                    Year.parse(potentialEID.substring(1, 3)),
                    chain(year =>
                        pipe(
                            SerialNumber.parse(potentialEID.substring(3, 6)),
                            chain(serialNumber => {
                                const eid = new EID(sex, year, serialNumber);
                                return eid.checkKey(potentialEID.substring(6))
                                    ? right(eid)
                                    : left(new ParsingException('Invalid key'));
                            })
                        )
                    )
                )
            )
        );
    }

    private checkKey(potentialKey: string): boolean {
        const key = parseInt(potentialKey);
        return !isNaN(key) && this.key() === key;
    }

    private stringWithoutKey(): string {
        return `${this.sex}${this.year}${this.serialNumber}`;
    }

    private toLong(): number {
        return parseInt(this.sex.toString() + this.year + this.serialNumber.toString());
    }

    key(): number {
        return 97 - (this.toLong() % 97);
    }

    toString(): string {
        return this.stringWithoutKey() + this.key().toString().padStart(2, '0');
    }

    equals(other: EID): boolean {
        if (!(other instanceof EID)) {
            return false;
        }
        return this.sex === other.sex &&
            this.year.equals(other.year) &&
            this.serialNumber.equals(other.serialNumber);
    }
}

File: exercise/C#/day22/EID/EID.cs
using LanguageExt;

namespace EID
{
    public readonly struct EID : IEquatable<EID>
    {
        private const int ValidLength = 8;

        private readonly Sex _sex;
        private readonly Year _year;
        private readonly SerialNumber _serialNumber;

        public EID(Sex sex, Year year, SerialNumber serialNumber)
        {
            _sex = sex;
            _year = year;
            _serialNumber = serialNumber;
        }

        // Override the default constructor for readonly struct generated at compile time...
        public EID()
        {
            _sex = Sex.Sloubi;
            _year = (Year) 1;
            _serialNumber = (SerialNumber) 1;
        }

        public static Either<ParsingError, EID> Parse(string potentialEID)
            => (from sex in SexParser.Parse(potentialEID[0])
                    from year in Year.Parse(potentialEID[1..3])
                    from serialNumber in SerialNumber.Parse(potentialEID[3..6])
                    select new {EID = new EID(sex, year, serialNumber), Key = potentialEID[6..8]})
                .Bind(parsedEID => CheckKey(parsedEID.Key, parsedEID.EID));

        private static Either<ParsingError, EID> CheckKey(string potentialKey, EID eid)
            => potentialKey.ToInt()
                .Filter(parsedKey => parsedKey == eid.Key())
                .Map(_ => eid)
                .ToEither(new ParsingError("invalid key"));

        private string StringWithoutKey() => $"{(int) _sex}{_year}{_serialNumber}";

        private int Key()
            => StringWithoutKey()
                .ToInt()
                .Map(x => 97 - x % 97)
                .IfNone(0);

        public bool Equals(EID other) =>
            _sex == other._sex && _year == other._year && _serialNumber == other._serialNumber;

        public override bool Equals(object? obj) => obj is EID other && Equals(other);

        public override int GetHashCode() =>
            HashCode.Combine((int) _sex, _year, _serialNumber);

        public override string ToString() => StringWithoutKey() + $"{Key():D2}";

        public static bool operator ==(EID left, EID right) => left.Equals(right);

        public static bool operator !=(EID left, EID right) => !(left == right);
    }
}